[Compose 페이먼츠 2단계] 미플 미션 제출합니다

HamBeomJoon opened the pull request:
# 셀프 체크리스트

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [x] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

- Slot 패턴이라는 것을 보고 CardListScreen에 적용을 해보려고 했는데, 잘 된 건지 궁금합니다. 현재 card가 isEmpty일때와 아닐때를 컴포저블 함수로 나눠놓긴 했는데, 그거까지 그냥 하나의 컴포저블 함수로 묶어야 하는 건지 의문이 듭니다.
CardListScreen에 content: @Composable (List<CardUiModel>) -> Unit로 카드가 1개이든 아니든 content 하나로 하려했는데, 그러면 Activity에서 ListScreen을 호출할 때 분기처리를 해서 선언해줘야 해 실패해서 이정도로만 나눠놓은 상태입니다.

- domain model을 value class로 만들어서 그런지 saver를 직접 구현하는 것도 코드가 너무 길고 더러워지는 것 같아 그냥 viewModel을 다시 뺐습니다.

- 화면에 출력되는 카드번호(* 마스킹이 포함된), 카드 소유자, 만료일에 대한 UiModel을 만들고 formatting을 진행했는데, UiModel을 만드는 것에 대해 어떻게 생각하나요??

## 참고 자료


## 스크린샷

https://github.com/user-attachments/assets/9dd9e9bf-db57-4594-9640-1ddd90683d37

HamBeomJoon committed:
docs: 2단계 기능 요구 사항 정의

HamBeomJoon committed:
chore: screen 패키지 분리

HamBeomJoon committed:
feat: 카드 리스트 화면 추가

- 카드 등록 Text 추가
- 카드 추가 Box 추가

HamBeomJoon committed:
chore: ktlint 적용

HamBeomJoon committed:
feat: 카드 목록 상태에 따른 UI 변경사항 적용

HamBeomJoon committed:
refactor: 카드 소유자 20자 제한으로 변경

HamBeomJoon committed:
feat: CardUiModel Parcelable 추가

HamBeomJoon committed:
feat: Card, CardUiModel mapper 추가

HamBeomJoon committed:
feat: PaymentCard에 입력값 (카드번호, 만료일, 소유자) 표시

HamBeomJoon committed:
feat: 카드 목록 화면에 추가한 카드들 UI 표시

HamBeomJoon committed:
feat: 새로 추가한 카드 저장 시 UiModel Intent로 넘기고 받기

HamBeomJoon committed:
feat: 새로운 카드 추가시 Toast 메시지 출력

HamBeomJoon committed:
refactor: Screen 매개변수로 activity 직접 넘기는 대신 lambda callback 넘기기

HamBeomJoon committed:
test: performTextInput으로 실제 텍스트 입력하듯이 테스트

HamBeomJoon committed:
test: 초기 setContent Before 함수로 이동

HamBeomJoon committed:
test: 카드 소유자 입력 테스트 수정

HamBeomJoon committed:
test: semantics에 contentDescription 추가, 테스트

HamBeomJoon committed:
refactor: CardListScreen 컴포저블 함수 분리

HamBeomJoon committed:
refactor: passwordVisualTransformation remember 변수로 관리

HamBeomJoon committed:
feat: 화면 회전 시 데이터 및 유효성 체크 상태 유지되도록 Saver 구현

HamBeomJoon committed:
feat: 화면 회전 시 Scroll되게 변경

HamBeomJoon committed:
refactor: viewModel 제거

HamBeomJoon committed:
refactor: textField 유효한지 변수로 관리

HamBeomJoon committed:
feat: 카드 0개, 카드 1개, 카드 2개일 때 Preview 추가

HamBeomJoon committed:
test: 카드 0개, 카드 1개, 카드 2개일 때 CardListScreen Test

HamBeomJoon committed:
docs: 프로그래밍 요구 사항 체크

HamBeomJoon committed:
refactor: filteredText 빈 값 null로 전달

HamBeomJoon committed:
refactor: 카드 목록 TopBar 추가 텍스트 contentDescription 수정

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/CardListTopBar.kt:
```diff
@@ -0,0 +1,55 @@
+                        modifier =
+                            modifier.semantics {
+                                this.contentDescription = "카드 목록 상단 추가 텍스트"
+                            },
```
1. `text`와 `contentDescription`이 하드코딩 되어있어요!
2. 코드 래빗이 알려준 modifier 관련 내용 `1단계 피드백 - Modifier - Don't` 부분 한 번 살펴보면 좋을 것 같아요 :)

HamBeomJoon replied:
1. 상수화를 진행하였습니다.

2. "외부에서 받은 modifier는 보통 가장 바깥 래퍼에만 적용하고, 내부 자식에는 별도의 Modifier를 사용하라." 인 것 같아서 매개변수로 받은 modifier를 그대로 쓰는 것을 수정하였습니다. 
32번째 줄에 테스트를 위한 semantics을 추가하는 것은 UI와 관계없는 부분 같은데 이부분도 Modifier를 새롭게 정의해서 사용하는게 좋다고 생각하나요? 아니면 상위에서 받은 modifier를 그대로 써도 된다고 생각하나요?

moondev03 replied:
> 32번째 줄에 테스트를 위한 semantics을 추가하는 것은 UI와 관계없는 부분 같은데 이부분도 Modifier를 새롭게 정의해서 사용하는게 좋다고 생각하나요? 아니면 상위에서 받은 modifier를 그대로 써도 된다고 생각하나요?

`semantics { contentDescription = "" }`을 사용하는게 왜 UI와 관계가 없는 것 같다고 생각하시는 지 궁금합니다.

그리고 부모에게서 받은 `modifer`를 여러 곳에서 사용하게 되면 예기치 못한 문제가 발생할 수 있다고 생각합니다.
`CenterAlignedTopAppBar`에 사용자가 원하는 `modifier`를 적용하고 싶어서 여러 메서드들을 적용하고 인자로 넣어준 경우
이 `modifier`는 `CenterAlignedTopAppBar`만을 생각하고 메서드들을 적용했을텐데 `Text`에도 적용되는 예기치 않은 문제가 발생할 것 같아요.

HamBeomJoon committed:
feat: 화면 회전 시 저장된 카드 리스트 정보 유지

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardScreen.kt:
```diff
@@ -0,0 +1,171 @@
+            restore = { Password(it) },
+        )
+
+    var cardNumber by rememberSaveable(stateSaver = cardNumberSaver) { mutableStateOf(null) }
+    var expired by rememberSaveable(stateSaver = expiredSaver) { mutableStateOf(null) }
+    var cardOwner by rememberSaveable(stateSaver = cardOwnerSaver) { mutableStateOf(CardOwner("")) }
+    var password by rememberSaveable(stateSaver = passwordSaver) { mutableStateOf(null) }
+    var showValidationError by rememberSaveable { mutableStateOf(false) }
+
+    val scrollState = rememberScrollState()
+
+    val isFormValid by remember(cardNumber, expired, cardOwner, password) {
+        derivedStateOf {
+            (cardNumber?.isValid == true) &&
+                (expired?.isValid == true) &&
+                (cardOwner?.isValid != false) &&
+                (password?.isValid == true)
+        }
+    }
```
`AddCardUiState`같은 이름의 상태 홀더를 정의한다면 상태를 관리하기 더욱 편리해질 수 있을 것 같아요

공식 문서에서는 이런 상태 홀더를 사용할 때의 이점으로 다음과 같이 설명하고 있습니다.
```
간단한 UI: UI가 상태를 바인딩합니다.
유지관리: 상태 홀더에 정의된 로직을 UI 자체를 변경하지 않고도 반복할 수 있습니다.
테스트 가능성: UI 및 상태 생성 로직을 독립적으로 테스트할 수 있습니다.
가독성: 코드 리더가 UI 표시 코드와 UI 상태 생성 코드 간의 차이점을 명확하게 알아볼 수 있습니다.
```

HamBeomJoon replied:
AddCardUiState를 추가하여 상태 관리를 하도록 변경하였습니다.
2eb32dcbbe07090e87094e608da7f67eeb082ae7

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardScreen.kt:
```diff
@@ -0,0 +1,171 @@
+            (cardNumber?.isValid == true) &&
+                (expired?.isValid == true) &&
+                (cardOwner?.isValid != false) &&
+                (password?.isValid == true)
```
다른 상태들은 `isValid == true`로 비교를 하고 있는데 `cardOwner`는 `isValid != false`로 비교하는 이유가 있을까요?

HamBeomJoon replied:
어 그러네요. 수정했습니다

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardScreen.kt:
```diff
@@ -0,0 +1,171 @@
+                onSaveClick = {
+                    showValidationError = !isFormValid
+                    if (isFormValid) {
+                        val cardUiModel =
+                            Card(
+                                number = cardNumber,
+                                expired = expired,
+                                owner = cardOwner,
+                                password = password,
+                            ).toPresentation()
```
현재 뷰에서 도메인 객체를 생성한 뒤 다시 UiModel로 변환하는 방식은 조금 어색한 것 같아요.

대신 바로 UiModel을 생성하면 자연스럽고, 
UiModel에 Mapper에서 작성한 포맷 함수 등 역할도 부여할 수 있을 것 같습니다.

HamBeomJoon replied:
직접 생성하도록 변경했습니다.
d1467c0b79aa2603ec6ce516c98f85773107ce41

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardScreen.kt:
```diff
@@ -0,0 +1,171 @@
+                        expired = expired,
+                        owner = cardOwner,
+                        password = password,
+                    ).toPresentation(),
```
여기도 다소 어색한 것 같아요.
위에서 말씀드린 상태 홀더를 사용하면 내부에서 CardUiModel를 반환하도록 해볼 수도 있을 것 같아요.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListActivity.kt:
```diff
@@ -0,0 +1,63 @@
+import woowacourse.payments.ui.theme.AndroidpaymentsTheme
+
+class CardListActivity : ComponentActivity() {
+    private val cardList = mutableStateListOf<CardUiModel>()
```
`카드 목록`이 `CardListScreen` 외부에서 수정 가능한 형태로 존재하고 있어 보여요 😮

HamBeomJoon replied:
수정한 것 같은데, 반영이 안되었던 것 같아요!

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListScreen.kt:
```diff
@@ -0,0 +1,200 @@
+
+@Composable
+private fun CardListContent(
+    cards: List<CardUiModel>,
+    modifier: Modifier = Modifier,
+    onAddCardClick: () -> Unit,
```
`modifier`를 `첫 번째 선택적 매개변수`로 하기를 권장하고 있습니다.
특별한 이유가 없다면 위치 조정 해주시면 좋을 것 같아요 :)

HamBeomJoon replied:
변경하였습니다.
eae8070d1ff2596439304ef591ce703125137859

moondev03 commented on app/src/main/res/values/strings.xml:
```diff
@@ -2,14 +2,14 @@
     <string name="card_owner_label">카드 소유자 이름(선택)</string>
     <string name="card_owner_placeholder">카드에 표시된 이름을 입력하세요.</string>
-    <string name="card_owner_length">%s / 30</string>
+    <string name="card_owner_length">%s / 20</string>
```
카드 소유자 이름 길이 요구사항은 다음과 같습니다.
```
카드 소유자 이름의 경우 입력 글자 제한이 30자이다.
```

HamBeomJoon replied:
카드 소유자가 너무 길어지면 ui가 2줄로 되는 등 예쁘지 않아 일단 20자로 해놨습니다

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListScreen.kt:
```diff
@@ -0,0 +1,200 @@
+@Composable
+private fun EmptyCardList(
+    modifier: Modifier = Modifier,
+    onAddCardClick: () -> Unit,
```
`modifier`를 `첫 번째 선택적 매개변수`로 하기를 권장하고 있습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListScreen.kt:
```diff
@@ -0,0 +1,200 @@
+            navigateToAddCard = { },
+        )
+    }
+}
```
케이스가 늘어날수록 Preview가 늘어나게 될 것 같아요.
`@PreviewParameter`를 활용해보면 좋을 것 같습니다 :)

HamBeomJoon replied:
오 신기한 방법이네요!👍

998f085ea6c1af6ee411239f716c0d9337fced31

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardScreen.kt:
```diff
@@ -0,0 +1,171 @@
+                password = password,
+                onPasswordChange = { password = it },
+                modifier = Modifier.fillMaxWidth(0.5f),
+                showValidationError = showValidationError,
```
모든 TextField가 showValidateionError하는 상태를 동일하게 사용하고 있는 것 같아요
이 경우 사용자는 어떤 입력이 잘못되었는지를 파악하기 어려울 것 같습니다.

HamBeomJoon replied:
잘못 된 textField는 테두리가 빨갛게되는데 파악하기 어렵다는게 구체적으로 얘기해줄 수 있을까요?

HamBeomJoon committed:
feat: 화면 회전 시 저장된 카드 리스트 정보 유지

HamBeomJoon committed:
refactor: 토스트 메시지 상수화, Intents Key 상수화

moondev03 reviewed (changes requested):
2단계 고생하셨습니다 👍 

아래 추가로 남긴 코멘트도 확인 부탁드려요

moondev03 commented:
전체적으로 문자열이 하드 코딩 되어있는 점이 아쉬웠습니다.
strings.xml 활용해주세요!

---

> Slot 패턴이라는 것을 보고 CardListScreen에 적용을 해보려고 했는데, 잘 된 건지 궁금합니다. 현재 card가 isEmpty일때와 아닐때를 컴포저블 함수로 나눠놓긴 했는데, 그거까지 그냥 하나의 컴포저블 함수로 묶어야 하는 건지 의문이 듭니다.
CardListScreen에 `content: @Composable (List) -> Unit`로 카드가 1개이든 아니든 content 하나로 하려했는데, 그러면 Activity에서 ListScreen을 호출할 때 분기처리를 해서 선언해줘야 해 실패해서 이정도로만 나눠놓은 상태입니다.

공식 문서에서 제공되는 [슬롯 기반 레이아웃](https://developer.android.com/develop/ui/compose/layouts/basics?hl=ko#slot-based-layouts)에 대한 설명입니다.

제가 이해한 슬롯 기반 레이아웃은 부모 Composable이 레이아웃의 위치와 구조만 정의하고, 실제 content는 외부에서 주입받아 보여주는 패턴입니다. 공식 문서 예제의 Column이나 Row처럼 내부 구조는 고정되어 있지만, 어떤 Composable을 넣을지는 호출 시점에 결정할 수 있는 형태입니다.

그래서 슬롯 기반 레이아웃에서는 내부 content가 카드가 비었는지 여부를 알 필요가 없고, 단순히 주어진 content만 보여주면 된다고 생각해요.
하지만 제 기준으로 현재 코드에서는 슬롯 기반 레이아웃의 형태가 보이지 않는 것 같습니다.

만약 Slot 패턴을 적용한다면, 아래와 같이 구현할 수 있을 것 같아요.
제가 이해한 바로 작성한 코드이니 참고용으로만 봐주세요 :)

```kotlin
@Composable
private fun CardListContent(
    modifier: Modifier = Modifier,
    content: @Composable () -> Unit,
) {
    val scrollState = rememberScrollState()

    Column(
        modifier =
            modifier
                .fillMaxSize()
                .padding(horizontal = 16.dp)
                .verticalScroll(scrollState),
        horizontalAlignment = Alignment.CenterHorizontally,
    ) {
        content()
        Spacer(modifier = Modifier.height(40.dp))
    }
}
```
```kotlin
Scaffold(
        topBar = {
            CardListTopBar(
                showAddButton = showAddButtonInTopBar,
                onAddClick = navigateToAddCard,
            )
        },
    ) { innerPadding ->
        CardListContent(modifier = Modifier.padding(innerPadding)) {
            if (cards.isEmpty()) {
                Spacer(modifier = Modifier.height(32.dp))
                Text(
                    text = "새로운 카드를 등록해주세요",
                    fontSize = 18.sp,
                    fontWeight = FontWeight.Bold,
                )
            }

            for (card in cards) {
                Spacer(modifier = Modifier.height(32.dp))
                PaymentCard(card = card)
            }

            if (cards.size <= 1) {
                Spacer(modifier = Modifier.height(32.dp))
                AddCardBox(onClick = navigateToAddCard)
            }
        }
    }
```

<br>

> domain model을 value class로 만들어서 그런지 saver를 직접 구현하는 것도 코드가 너무 길고 더러워지는 것 같아 그냥 viewModel을 다시 뺐습니다.

코멘트에서 작성한 내용인데 카드의 모든 상태들을 따라 작성하는게 아닌 UiState로 묶어서 하나로 관리하게 된다면 여러가지 측면에서 해결되는 점이 많을 것 같아요.

자세한 내용은 [해당 문서](https://developer.android.com/topic/architecture/ui-layer/stateholders?hl=ko) 참고해보면 좋을 것 같습니다 :)

<br>

> 화면에 출력되는 카드번호(* 마스킹이 포함된), 카드 소유자, 만료일에 대한 UiModel을 만들고 formatting을 진행했는데, UiModel을 만드는 것에 대해 어떻게 생각하나요??

저 역시 화면에 출력되는 UiModel을 만들고 그 안에서 포맷팅을 처리했습니다.

그런데 미플의 코드를 보니 매번 Card라는 도메인 객체를 만든 뒤 toPresentation()을 거쳐 포맷팅된 값을 기반으로 다시 UiModel을 생성하고 있는 것 같아요. 이렇게 되면 뷰에서 도메인 객체와 UiModel을 동시에 다루게 되어 다소 혼란스러운 구조가 될 수 있다고 생각합니다.

차라리 UiModel 자체가 포맷팅된 값을 제공하도록 설계하면, 뷰에서는 UiModel만 바라보면 되기 때문에 좀 더 단순하고 유연한 구조가 될 것 같아요.

HamBeomJoon committed:
refactor: modifier 하위로 전파 방지

HamBeomJoon committed:
refactor: modifier 매개변수 첫번째로 이동

HamBeomJoon committed:
refactor: strings 상수화

HamBeomJoon committed:
refactor: PreviewParameter로 프리뷰 분리

HamBeomJoon committed:
refactor: 카드 소유자 입력 필드 및 카드 추가 화면 UI 상태 개선

- CardOwnerInputField의 cardOwner 매개변수를 nullable에서 non-nullable로 변경하고, 관련 로직을 수정했습니다.
- AddCardUiState 데이터 클래스를 추가하여 카드 추가 화면의 UI 상태를 관리하도록 했습니다.
- AddCardScreen에서 개별 상태 변수 대신 AddCardUiState를 사용하도록 리팩터링하고, 상태 저장 로직을 Saver를 사용하여 개선했습니다.

HamBeomJoon committed:
refactor: card 도메인 모델 uiModel로 변환하지 않고 직접 UiModel 생성하도록 수정

HamBeomJoon committed:
refactor: ui 유틸리티 패키지 분리 및 번들 키 상수화

HamBeomJoon committed:
refactor: cardListContent slot 패턴 적용

HamBeomJoon committed:
chore: ktlint 적용

HamBeomJoon commented:
slot 패턴을 적용해 보았고, 전반적으로 strings로 분리를 진행하였습니다.
또한, UiState를 도입해 상태관리를 하도록 변경했습니다. 

- contentDescription같은 경우는 text와 다르게 stringResources로 안되던데, 그냥 다이스가 작성한 것처럼 테스트코드에서 fake나 mock 객체를 만들고 testTag를 다는 것이 낫다고 생각하시나요?

https://github.com/woowacourse/android-payments/pull/31#discussion_r2331824520
여기서 남겨주신 것 처럼  테스트를 위한 semantics을 추가하는 것은 UI와 관계없는 부분 같은데 이부분도 Modifier를 새롭게 정의해서 사용하는게 좋다고 생각하나요? 아니면 상위에서 받은 modifier를 그대로 써도 된다고 생각하나요?

HamBeomJoon committed:
refactor: 파라미터 순서 조정

HamBeomJoon committed:
refactor: Activity에서 LocalContext 대신 this로 컨텍스트 사용

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/PasswordInputField.kt:
```diff
@@ -26,31 +29,34 @@ import woowacourse.payments.ui.theme.AndroidpaymentsTheme
     modifier: Modifier = Modifier,
+    password: Password?,
+    onPasswordChange: (Password?) -> Unit,
     showValidationError: Boolean = false,
```
의도하신 순서인가요?

> 컴포저블 함수를 설계할 때 첫 번째 선택적(Optional) 매개변수로 Modifier를 받도록 설계하자.

[(docs)](https://android.googlesource.com/platform/frameworks/support/+/androidx-main/compose/docs/compose-api-guidelines.md#elements-accept-and-respect-a-modifier-parameter)

HamBeomJoon replied:
무조건 modifier가 맨 위로 가야한다고 착각했네요 😅

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardScreen.kt:
```diff
@@ -0,0 +1,161 @@
+                    showValidationError = bundle.getBoolean(VALIDATION_ERROR_KEY),
+                )
+            },
+        )
```
`rememberSaver`를 사용하기 위한 방법 중 Parcelable이나 Serializable를 사용하는 방법과 미플이 사용한 Saver를 구현하는 방법 중 후자를 선택하신 이유가 궁금합니다!

HamBeomJoon replied:
AddCardUiState가 Parcelable을 사용하고 있지 않아 위 방식으로 구현을 했습니다!

moondev03 commented on app/src/androidTest/java/woowacourse/payments/CardListScreenTest.kt:
```diff
@@ -0,0 +1,86 @@
+        composeTestRule.onNodeWithText(card1.number).assertIsDisplayed()
+        composeTestRule.onNodeWithText(card1.expired).assertIsDisplayed()
+        composeTestRule.onNodeWithText(card1.owner).assertIsDisplayed()
+        composeTestRule.onNodeWithContentDescription("카드 추가").assertIsDisplayed()
```
`assertAll`를 활용하면 모든 케이스가 수행되도록 할 수 있을 것 같아요!

HamBeomJoon replied:
오 좋은 방법 감사합니다

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/CardNumberInputField.kt:
```diff
@@ -16,9 +18,9 @@ import woowacourse.payments.ui.theme.AndroidpaymentsTheme
 
 @Composable
 fun CardNumberInputField(
+    modifier: Modifier = Modifier,
```
원래 Compose Guideline대로 잘 하고 계셨는데 전체적으로 `modifier` 매개변수 위치가 최상위로 변경된 것 같아요
이유가 있을까요?

HamBeomJoon replied:
다시 수정했습니다!

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListScreen.kt:
```diff
@@ -0,0 +1,160 @@
+        content()
+        Spacer(modifier = Modifier.height(40.dp))
+    }
+}
```
스크롤이나 최하단 Spacer의 경우 `카드가 없는 케이스`와 `카드가 1장인 케이스`의 경우에도 필요할까요?
저는 필요하지 않기 때문에 2장 이상인 경우에만 적용될 수 있는 구조를 채택했어요.
미플은 어떻게 생각하시나요?

HamBeomJoon replied:
변수 하나를 상위에 더 만들어서 분기처리를 통해 scrollState적용, 하단 Spacer 적용을 했는데 코드가 늘어나는 것 같아 애매하네요..

하지만 지금 코드는 지금 미션이니까 돌아가는 UI같긴해서 변경했습니다.

643556a93e0952a9ee3cae0eede0e46222c69451

moondev03 commented on app/src/main/java/woowacourse/payments/ui/AddCardUiState.kt:
```diff
@@ -0,0 +1,19 @@
+    val expired: Expired? = null,
+    val cardOwner: CardOwner = CardOwner(""),
+    val password: Password? = null,
+    val showValidationError: Boolean = false,
```
`showValidationError`는 어떤 카드가 유효하지 않은 지 알기 힘들 것 같아요.
각 모델의 isValid와 조합하여 사용하고는 있지만 만약 예외로 보여줘야 하는 케이스가 많아질 경우 복잡합 구조를 가지게 될 수 있을 것 같습니다.

미플은 어떻게 생각하시나요?

HamBeomJoon replied:
AddCardError enum Class를 생성해 간단하게(?) 에러 상태를 하나씩 만들고 해당 상태에 따라 처리했습니다.
56f8dc3a017380bf33516880de27047fd33cdd9e

moondev03 commented on app/src/main/java/woowacourse/payments/domain/Card.kt:
```diff
@@ -0,0 +1,8 @@
+    val expired: Expired?,
+    val owner: CardOwner?,
+    val password: Password?,
+)
```
저는 도메인 모델에는 비논리적인 오류가 포함되면 안 된다고 생각합니다.
그래서 현재 Card가 모든 카드 정보를 nullable로 가지고 있는 것이 어색하게 느껴져요.
카드에 모든 카드 정보가 비어 있는 경우가 과연 있을까요?

HamBeomJoon replied:
그러네요. 왜 nullable로 했는지 기억이 안나네요 🥲

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListScreen.kt:
```diff
@@ -0,0 +1,160 @@
+            navigateToAddCard = { },
+        )
+    }
+}
```
`EMPTY / SIGNLE / MULTIPLE`로 Preview를 정의해두면 좋을 것 같아요

moondev03 commented on app/src/main/java/woowacourse/payments/ui/util/CardMapper.kt:
```diff
@@ -0,0 +1,24 @@
+fun formatExpired(expired: Expired?): String {
+    if (expired == null || expired.value.isEmpty()) return ""
+    return expired.value.chunked(2).joinToString(" / ")
+}
```
특정 모델을 문자열로 포맷팅하는 작업이 Util의 책임일까요?

HamBeomJoon replied:
그러네요 지금 UiModel에서만 사용중이어서 해당 파일 안으로 이동시켰습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/AddCardUiState.kt:
```diff
@@ -0,0 +1,19 @@
+import woowacourse.payments.domain.Expired
+import woowacourse.payments.domain.Password
+
+data class AddCardUiState(
```
해당 UiState 파일의 위치는 `addCard` 패키지 아래에 있는게 적절하다고 생각해요!

HamBeomJoon replied:
옮겼습니다

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListScreen.kt:
```diff
@@ -0,0 +1,160 @@
+
+@Composable
+fun CardListScreen(
+    cards: List<CardUiModel>,
```
`AddCardScreen`에서는 UiState를 활용했는데 여기서는 사용하지 않은 특별한 이유가 있을까요?
여기서도 사용한다면 어제 강의에서 다룬 `State Holder`의 이점을 얻을 수 있을 것 같아요

HamBeomJoon replied:
UiState, StateHolder를 사용해보도록 수정했습니다
400cd2be6796364884f140afb7bf1c3bae93323a

moondev03 reviewed (changes requested):
리뷰 반영 고생하셨습니다 👍 
궁금한 점 코멘트로 남겼으니 확인 부탁드려요 :)

다음 요청 때는 카드 등록 화면에 대한 테스트 코드도 작성되면 좋을 것 같아요!

---

> contentDescription같은 경우는 text와 다르게 stringResources로 안되던데, 그냥 다이스가 작성한 것처럼 테스트코드에서 fake나 mock 객체를 만들고 testTag를 다는 것이 낫다고 생각하시나요?

저는 프로덕션 코드에 테스트만을 위한 `testTag`가 포함되는 것이 적절하지 않다고 판단되면 `contentDescription`을 사용하고
굳이 코드가 오염된다고 생각하지 않는다고 판단되면 `testTag`를 사용해도 괜찮다 생각합니다.

다만, `contentDescription`은 테스트만을 위한 방법이 아니고 접근성을 위함이라는 근거도 있다고 생각해서 지금은 `contentDescription`이 더 적절한 것 같다는 생각입니다!

<br>

> 여기서 남겨주신 것 처럼 테스트를 위한 semantics을 추가하는 것은 UI와 관계없는 부분 같은데 이부분도 Modifier를 새롭게 정의해서 사용하는게 좋다고 생각하나요? 아니면 상위에서 받은 modifier를 그대로 써도 된다고 생각하나요?

semantics { contentDescription = "" }을 사용하는게 왜 UI와 관계가 없는 것 같다고 생각하시는 지 궁금합니다.

그리고 부모에게서 받은 modifer를 여러 곳에서 사용하게 되면 예기치 못한 문제가 발생할 수 있다고 생각합니다.
CenterAlignedTopAppBar에 사용자가 원하는 modifier를 적용하고 싶어서 여러 메서드들을 적용하고 인자로 넣어준 경우
이 modifier는 CenterAlignedTopAppBar만을 생각하고 메서드들을 적용했을텐데 Text에도 적용되는 예기치 않은 문제가 발생할 것 같아요.

HamBeomJoon committed:
refactor: Card 도메인 모델 필드 nullable하지 않게 변경

HamBeomJoon committed:
refactor: CardUiModel에 카드 번호 및 만료일 포맷팅 로직 이동

- CardUiModel에 카드 번호 및 만료일을 포맷팅하는 로직을 추가했습니다.
- 입력 필드 컴포저블에서 nullable 파라미터를 제거하고 기본값을 설정했습니다.
- 입력 필드 컴포저블에서 유효성 검사 로직을 수정했습니다.

HamBeomJoon committed:
refactor: `AddCardUiState`를 `addCard` 패키지로 이동

HamBeomJoon committed:
refactor: CardMapper.kt 삭제 및 CardUiModel.kt로 통합

HamBeomJoon committed:
test: CardListScreenTest assertAll로 수정

HamBeomJoon committed:
feat: 카드 1개일 경우 스크롤 및 하단 여백 제거

HamBeomJoon committed:
refactor: AddCardScreen의 상태 관리 로직을 AddCardStateHolder로 분리

HamBeomJoon committed:
refactor: `CardUiModel` 패키지 이동 및 적용

HamBeomJoon committed:
feat: `CardListScreen` 상태 관리 로직 `CardListStateHolder`로 분리

HamBeomJoon committed:
feat: 입력값 유효성 검증 오류 메시지 표시

HamBeomJoon committed:
test: UI Test 패키지 분리

HamBeomJoon committed:
test: 카드 추가 화면 UI Test 추가

HamBeomJoon commented:
크게 변경된 사항은 아래와 같습니다. UiState와 StateHolder의 차이?, 역할을 어디까지 해야하는지 대해 고민해봤는데..  다이스의 의견도 궁금합니다.!

- UiState + StateHolder를 도입해보았습니다. 
- Input TextField에서 에러 상태별로 enum class를 추가해 메시지를 표시하는 식으로 관리하도록 변경했습니다.
- 카드 추가 화면 테스트도 진행하였습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/domain/Card.kt:
```diff
@@ -0,0 +1,8 @@
+package woowacourse.payments.domain
+
+data class Card(
+    val number: CardNumber,
+    val expired: Expired,
+    val owner: CardOwner,
+    val password: Password,
+)
```
사용하고 있지 않은 것 같아요

moondev03 replied:
저는 도메인 모델이 생성되었다는 것은 이미 논리적인 케이스가 배제되었다는 의미라고 생각합니다.
그런 관점에서 보면 현재 미플이 구현한 모델은 도메인 모델이라고 보기 어렵다고 생각해요.
이 부분에 대해 미플의 의견이 궁금합니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/CardNumberInputField.kt:
```diff
@@ -8,41 +10,64 @@ import androidx.compose.runtime.remember
-    cardNumber: CardNumber?,
-    onCardNumberChange: (CardNumber?) -> Unit,
     modifier: Modifier = Modifier,
-    showValidationError: Boolean = false,
+    cardNumber: CardNumber? = null,
+    onCardNumberChange: (CardNumber) -> Unit,
+    error: AddCardError? = null,
```
매개변수 순서가 이상한 것 같아요!

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/CardNumberInputField.kt:
```diff
@@ -8,41 +10,64 @@ import androidx.compose.runtime.remember
+            },
+            modifier =
+                modifier.semantics {
+                    this.contentDescription = "Card Number Input Field"
```
`contentDescription`을 하드 코딩 하신 이유가 있나요?

해당 문자열도 다른 문자열과 동일하게 사용자에게 노출되는 값이기에 문자열 리소스로 관리하는게 적절하다고 생각해요!

테스트만을 위해 추가하셨다면 testTag가 더 적절하다고 생각합니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/CardNumberInputField.kt:
```diff
@@ -8,41 +10,64 @@ import androidx.compose.runtime.remember
+
+        error?.let {
             Text(
-                text = stringResource(R.string.card_number_placeholder),
-                color = Color.LightGray,
+                text = stringResource(R.string.card_number_invalid),
```
에러 메시지를 내부에서 처리하는 것과 외부에서 넣어주는 것 중 어느 것이 적절하다고 생각하시나요?

HamBeomJoon replied:
enum class에 메시지까지 넣도록 수정하였습니다

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/CardNumberInputField.kt:
```diff
@@ -55,3 +80,15 @@ fun CardNumberInputPreview() {
     }
 }
+
+@Composable
+@Preview(showBackground = true)
+fun CardNumberInputErrorPreview() {
+    AndroidpaymentsTheme {
+        CardNumberInputField(
+            cardNumber = CardNumber(""),
+            onCardNumberChange = { },
+            error = AddCardError.CARD_NUMBER_INVALID,
+        )
+    }
+}
```
에러가 일어나지 않은 경우도 보여줄 수 있을 것 같아요

HamBeomJoon replied:
바로 위에 있습니다 😅

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/CardOwnerInputField.kt:
```diff
@@ -8,59 +9,92 @@ import androidx.compose.runtime.Composable
-    cardOwner: CardOwner?,
-    onOwnerChange: (CardOwner?) -> Unit,
     modifier: Modifier = Modifier,
-    showValidationError: Boolean = false,
+    cardOwner: CardOwner? = null,
+    onOwnerChange: (CardOwner) -> Unit,
+    error: AddCardError? = null,
```
여기도 매개면수 순서 수정이 필요해 보입니다

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/CardOwnerInputField.kt:
```diff
@@ -8,59 +9,92 @@ import androidx.compose.runtime.Composable
-                    ),
-                modifier = Modifier.fillMaxWidth(),
-                color = Color.Gray,
-                style = MaterialTheme.typography.bodyMedium,
-                textAlign = TextAlign.End,
+                text = stringResource(R.string.card_owner_invalid),
+                modifier =
+                    Modifier.semantics {
+                        this.contentDescription = "Card Owner Input Error"
+                    },
+                color = Color.Red,
+                fontSize = 12.sp,
             )
```
`OutlinedTextField`의 `supportingText`을 활용해보는 것도 좋을 것 같아요

HamBeomJoon replied:
헉 그러네요.. 왜 소유자에서 글자수 표시에선 사용하고, 에러메시지에서는 생각을 못했네요.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/NewCardTopBar.kt:
```diff
@@ -19,20 +19,13 @@ import woowacourse.payments.ui.theme.AndroidpaymentsTheme
 @OptIn(ExperimentalMaterial3Api::class)
 @Composable
 fun NewCardTopBar(
+    modifier: Modifier = Modifier,
     onBackClick: () -> Unit,
     onSaveClick: () -> Unit,
```
`modifier`는 선택 가능한 매개변수 중 가장 위로 배치하면 좋을 것 같아요

moondev03 commented on app/src/main/java/woowacourse/payments/ui/model/CardUiModel.kt:
```diff
@@ -0,0 +1,28 @@
+    val formattedExpired get() = formatExpired(expired)
+}
+
+fun formatCardNumber(cardNumber: String): String {
+    val front = cardNumber.take(8).chunked(4).joinToString(" - ")
+
+    return if (cardNumber.length <= 8) {
+        front
+    } else {
+        val maskedCount = (cardNumber.length - 8).coerceAtLeast(0)
+        val masked = "*".repeat(maskedCount).chunked(4).joinToString(" - ")
+        "$front - $masked"
+    }
+}
+
+fun formatExpired(expired: String): String = expired.chunked(2).joinToString(" / ")
```
`CardUiModel` 외부에 함수를 작성하신 이유가 있나요?
저는 내부에 존재하는게 더 적절한 위치인 것 같다고 생각해요

HamBeomJoon replied:
내부로 이동시켰습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardStateHolder.kt:
```diff
@@ -0,0 +1,76 @@
+    var uiState by mutableStateOf(initialState)
+        private set
+
+    fun updateCardNumber(newNumber: CardNumber?) {
+        uiState = uiState.copy(cardNumber = newNumber)
+    }
+
+    fun updateExpired(newExpired: Expired?) {
+        uiState = uiState.copy(expired = newExpired)
+    }
+
+    fun updateCardOwner(newOwner: CardOwner) {
+        uiState = uiState.copy(cardOwner = newOwner)
+    }
+
+    fun updatePassword(newPassword: Password?) {
+        uiState = uiState.copy(password = newPassword)
+    }
```
각 정보들을 `nullable`로 받는 이유는 무엇인가요?

HamBeomJoon replied:
nullable하지 않게 변경했습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardUiState.kt:
```diff
@@ -0,0 +1,37 @@
+import woowacourse.payments.ui.model.CardUiModel
+
+data class AddCardUiState(
+    val cardNumber: CardNumber? = null,
+    val expired: Expired? = null,
+    val cardOwner: CardOwner = CardOwner(""),
+    val password: Password? = null,
```
각 정보들이 `Null`인 경우가 언제인가요?
지금 구조를 보면 `Null`일 경우가 없는 것 같아요

HamBeomJoon replied:
nullable하지 않게 변경했습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardUiState.kt:
```diff
@@ -0,0 +1,37 @@
+    val password: Password? = null,
+    val errors: Set<AddCardError> = emptySet(),
+) {
+    val isFormValid: Boolean get() = errors.isEmpty()
+    val cardNumberError: AddCardError? get() = errors.find { it == AddCardError.CARD_NUMBER_INVALID }
+    val expiredError: AddCardError? get() = errors.find { it == AddCardError.EXPIRED_INVALID }
+    val ownerError: AddCardError? get() = errors.find { it == AddCardError.OWNER_INVALID }
+    val passwordError: AddCardError? get() = errors.find { it == AddCardError.PASSWORD_INVALID }
```
이런 구조를 채택하면서 얻은 이점이 무엇인지 궁금합니다!

moondev03 replied:
`get()`이 필요하지 않은 것 같습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/addCard/AddCardUiState.kt:
```diff
@@ -0,0 +1,37 @@
+    val ownerError: AddCardError? get() = errors.find { it == AddCardError.OWNER_INVALID }
+    val passwordError: AddCardError? get() = errors.find { it == AddCardError.PASSWORD_INVALID }
+
+    fun validate(): AddCardUiState {
+        val newErrors = mutableSetOf<AddCardError>()
+        if (cardNumber?.isValid != true) newErrors.add(AddCardError.CARD_NUMBER_INVALID)
+        if (expired?.isValid != true) newErrors.add(AddCardError.EXPIRED_INVALID)
+        if (!cardOwner.isValid) newErrors.add(AddCardError.OWNER_INVALID)
+        if (password?.isValid != true) newErrors.add(AddCardError.PASSWORD_INVALID)
+        return copy(errors = newErrors)
+    }
```
함수로 분리하신 이유가 있으신가요?

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListScreen.kt:
```diff
@@ -0,0 +1,161 @@
+
+@Composable
+fun CardListScreen(
+    stateHolder: CardListStateHolder = remember { CardListStateHolder() },
+    navigateToAddCard: () -> Unit,
```
`stateHolder`는 기본 값이 있으니 `navigateToAddCard` 아래에 배치되어야 할 것 같아요

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListScreen.kt:
```diff
@@ -0,0 +1,161 @@
+            )
+        },
+    ) { innerPadding ->
+        CardListContent(
+            modifier = Modifier.padding(innerPadding),
+            enableScroll = uiState.enableScroll,
+        ) {
+            if (uiState.cards.isEmpty()) {
+                Spacer(modifier = Modifier.height(32.dp))
+                Text(
+                    text = stringResource(R.string.card_list_add_new_card),
+                    fontSize = 18.sp,
+                    fontWeight = FontWeight.Bold,
+                )
+            }
+
+            for (card in uiState.cards) {
+                Spacer(modifier = Modifier.height(32.dp))
+                PaymentCard(card = card)
+            }
+
+            if (uiState.cards.size <= 1) {
+                Spacer(modifier = Modifier.height(32.dp))
+                AddCardBox(onClick = navigateToAddCard)
+            }
+        }
+    }
```
만약 `uiState.cards.isEmpty`인 경우에만 다른 레이아웃을 배치해야 한다면 어떻게 해야 할까요?
제가 지난번 리뷰에서 말씀드린 의도는 상황별로 알맞은 레이아웃이 적용될 수 있도록 분리하는 것이었습니다.

HamBeomJoon replied:
Empty일때 EmptyCardLayout이 뜨도록 수정했습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListScreen.kt:
```diff
@@ -0,0 +1,161 @@
+}
+
+@Composable
+fun AddCardBox(onClick: () -> Unit) {
+    Box(
+        modifier =
+            Modifier
+                .size(width = 240.dp, height = 140.dp)
+                .clip(RoundedCornerShape(5.dp))
+                .background(Color.LightGray)
+                .clickable(onClick = onClick),
+        contentAlignment = Alignment.Center,
+    ) {
+        Icon(Icons.Default.Add, contentDescription = "카드 추가")
+    }
+}
```
`public`으로 열어둔 이유가 재사용이라면 파일을 분리하는게 좋을 것 같아요!

HamBeomJoon replied:
더 이상 재사용하진 않아서 private로 변경했습니다.

moondev03 commented on app/src/main/java/woowacourse/payments/ui/screen/cardList/CardListUiState.kt:
```diff
@@ -0,0 +1,10 @@
+data class CardListUiState(
+    val cards: List<CardUiModel> = emptyList(),
+) {
+    val showAddButton: Boolean get() = cards.size > 1
+    val enableScroll: Boolean get() = cards.size > 1
```
`get()`은 필요하지 않은 것 같아요.

`enableScroll`은 필요한 상태일까요?
스크롤 가능 여부는 cards.size만으로 결정되지는 않을 것 같아요

moondev03 commented on app/src/main/java/woowacourse/payments/ui/component/ExpiredInputField.kt:
```diff
@@ -8,41 +10,64 @@ import androidx.compose.runtime.remember
+                onExpiredChange(Expired(filteredText))
+            },
+            modifier =
+                modifier.semantics {
+                    this.contentDescription = "Expired Input Field"
+                },
```
매개변수로 받은 `modifier`는 최대한 가장 바깥쪽에만 적용되야 하는 이유에 대해서 리뷰를 남겼던 것 같아요.
혹시 기억나지 않으신다면 `LMS - 1단계 피드백 (Modifier - Don't)` 부분 참고해 주세요 :)

moondev03 reviewed (approved):
리뷰 반영 고생하셨습니다 👍 
궁금한 점 코멘트로 남겼으니 확인 부탁드려요 :)
